src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectManaged.cs

```diff
@@ -347,8 +382,26 @@ internal override uint GenerateSspiClientContext(byte[] receivedBuff, uint recei
 
         internal override uint WaitForSSLHandShakeToComplete(out int protocolVersion)
         {
-            protocolVersion = Handle.ProtocolVersion;
+            protocolVersion = GetSessionSNIHandleHandleOrThrow().ProtocolVersion;
             return 0;
         }
+
+        private SNIHandle GetSessionSNIHandleHandleOrThrow()
+        {
+            SNIHandle? sessionHandle = _sessionHandle;
+            if (sessionHandle is null)
+            {
+                ThrowClosedConnection();
+            }
+            return sessionHandle!;
```

Comment:
This non-null assertion is only needed on netstandard2.0, netcore 2.1 and later understand the flow correctly.